### ISSUE TITLE
Fields serialization (Long, Integer, Float, Double, etc.)

### ISSUE THREAD

**Issue by wowkin2** [elastic/elasticsearch-dsl-py#785]
Have similar issue like in [Problem with coerce? #21053](https://github.com/elastic/elasticsearch/issues/21053) during import from CSV
> I'm trying to make es cast some strings to integers and floats, but they seem to stay in as strings.

If I have field Double field and put there String with correct number "3.14" it will save me String in that Double type index. When I try to get it in Python code I'll get string without deserialization.
Only Boolean fields works correct (because it has redefined [_deserialize method](https://github.com/elastic/elasticsearch-dsl-py/blob/master/elasticsearch_dsl/field.py#L261))

Do you plan to add serialization for other fields or can I do that or this should be done on my application side?

**Comment by okke-formsma** [comment]
I've looked into this a bit.

There are three options for numeric datatypes that influence serialization: 'coerce', 'default', and 'ignore_malformed' (and combinations of these options). https://www.elastic.co/guide/en/elasticsearch/reference/current/number.html.

Should elasticsearch_dsl accept a string like 'nan' or 'inf', even though elasticsearch will fail on them? (https://github.com/elastic/elasticsearch-py/issues/486)

**Comment by 3lnc** [comment]
@okke-formsma 
I guess, big part of the problem is how `coerce` works in ES.

 - it does not coerce data in `_source`
 - its inconsistent, i.e. it raise error when trying to coerce malformed string to numeric, meanwhile silently losing data while truncating decimals to integers

And to be honest, I have no idea if its actually usable for someone in the way its implemented now.

**Comment by HonzaKral** [comment]
The reason why I originally didn't do serialization of numerical types is twofold - first I don't believe it should be needed when `json` fully supports those data types - there should never be any need or desire to express numeric values as strings for example.

Second reason is performance - currently we only coerce fields that absolutely require it and even as such it is a significant performance hit when instantiating a lot of objects, no matter what we do we need to make it optional to only be used when explicitly asked for.

**Comment by okke-formsma** [comment]
@3lnc, @HonzaKral, thanks for pitching in. To summarise your posts; elasticsearch_dsl should be transparent in its dealing with types - junk in is junk out.

Maybe we should not parse 'false' to `False` for a boolean field to stay consistent? (undoing #788)

**Comment by HonzaKral** [comment]
@okke-formsma that is a very good point about consistency with `"false"`, thank you for bringing it up!

The reason I implemented this is to be consistent with elasticsearch which will, for a declared boolean field, interpret that as `False` whereas python would make it a `True`. I will need to take a while and think about this.

I think there are some obvious things we need to be doing (and are) like deserializing datetimes and nested objects. I am, personally, not a fan of other (de)serializing other primitive types like numbers and booleans. I did make an exception for `"false"` because the behavior was wrong - we would turn it into `True` which is opposite of what elasticsearch is doing and not expected - we could be expected to leave it alone or coerce it same as elasticsearch would.

I am leaning towards not coercing the primitive types (int/float/boolean/...), optionally adding a flag to explicitly force the coercion. Question is whether we should validate the values and throw `ValidationError` when wrong type is present. I'd say yes (esp since users can opt out of validation).

What do you all think?

Thank you so much for this discussion, feedback is important and I really appreciate it!

**Comment by wowkin2** [comment]
It would be great to have at least optional flag to explicitly force the coercion.
In my case, I have web application that accept values from http requests and csv files. In both cases they provide string values that should be serialized somewhere. And if I have 50+ fields it is not good to convert them manually. If I create class that able to convert each field it will look the same as Elasticsearch-DSL model. So I expect that my model know about all field types and able to convert that automatically.
Currently, I overrided all Fields that I use with custom serialization:
```python
import elasticsearch_dsl


class Float(elasticsearch_dsl.Float):

    def _deserialize(self, data):
        if not data:
            return None
        return float(data)
```

**Comment by okke-formsma** [comment]
There are two places where deserialization may occur and I think it is helpful to consider these separately.

1) from python to elasticsearch_dsl
2) from elasticsearch (json) to elsaticsearch_dsl.

In the case of 1) there are many other ways to validate input; WTForms, Marshmallow, etc. It would be a duplication of effort to include this, and there are many edge cases it's hard to do right in all situations.

In the case of 2) elasticsearch_dsl is the bridge between elasticsearch and the python application. In my mind it makes sense to deserialize the data from elasticsearch to the correct python type, and serialize the python types correctly to elastic if the serialization/deserialization does not natively work because the JSON format does not support the type (IPAddress, datetime, etc).

What do you guys would the 'right' behavior be when the string '3' is stored on an Integer field? Consider both immediately after assignment to the Integer field, after storing in elasticsearch, and after retrieving from elasticsearch.

**Comment by HonzaKral** [comment]
@okke-formsma I like the distinction, which we currently don't have, of loading data from ES and from python assignment. If we split those code paths (by only doing deserialization in `from_es` method) it would make things much simpler, that is the way to go. Thank you!

@wowkin2 from what you are describing you should really clean your data in your http/csv handling where it belongs and not force your model layer (`DocType`) to know about any arbitrary formats that cane come through the wire. I agree that validation (and optional coercion) might be nice, but I don't think it is required at this point as there is a straightforward solution.

**Comment by wowkin2** [comment]
@HonzaKral, in that case I'll have 2 almost identical models one for `elasticsearch-dsl`, another for `marshmallow` library that do almost the same job. And if I or my colleague will want to make change - we need to don't forget to update it in both files, which definitely can lead to bugs.
Agree that this can work but don't want to have duplications.

**Comment by okke-formsma** [comment]
@HonzaKral, I'm comfortable improving the (de)serialize functions but don't feel comfortable in the codebase to do the es/python assignment split. I think deleting ObjectBase.__setattr__() would remove deserialization from the 'python path', but I don't know if that covers all cases.

**Comment by HonzaKral** [comment]
I implemented the split in 9574374 so we now only deserialize data when it comes from elasticsearch and not when assigning values from python. Thank you all for the discussion, it was super helpful!

@wowkin2 I understand your concerns about duplication - you can still use the `from_es` class method  which will trigger the deserialization. I will be adding more deserialization (most likely via #791) logic there so hopefully it will work for you as well from that direction. If it doesn't please feel free to get in touch and we will see what makes sense. Thank you for your patience!

**Comment by wowkin2** [comment]
Now looks good for me. Thank you!
